kernel/src/scan/partition_skipping.rs

```diff
@@ -0,0 +1,133 @@
+        visitor.visit_rows_of(partitions.as_ref())?;
+        Ok(visitor.selection_vector.clone())
+    }
+}
+
+struct PartitionVisitor {
+    pub(crate) selection_vector: Vec<bool>,
+    predicate: Arc<Expression>,
+    schema: SchemaRef,
+}
+
+impl PartitionVisitor {
+    pub(crate) fn new(predicate: &Arc<Expression>, schema: &SchemaRef) -> Self {
+        Self {
+            selection_vector: Vec::default(),
+            predicate: Arc::clone(predicate),
+            schema: Arc::clone(schema),
+        }
+    }
+}
+
+impl RowVisitor for PartitionVisitor {
+    fn selected_column_names_and_types(&self) -> (&'static [ColumnName], &'static [DataType]) {
+        static NAMES_AND_TYPES: LazyLock<crate::schema::ColumnNamesAndTypes> =
+            LazyLock::new(|| {
+                (
+                    vec![column_name!("output")],
+                    vec![DataType::Map(Box::new(MapType::new(
+                        DataType::STRING,
+                        DataType::STRING,
+                        true,
+                    )))],
+                )
+                    .into()
+            });
+        NAMES_AND_TYPES.as_ref()
+    }
+
+    fn visit<'a>(&mut self, row_count: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<()> {
+        let getter = getters[0];
+        for i in 0..row_count {
+            let val = getter.get_map(i, "output")?.map(|m| {
+                let partition_values = m.materialize();
+                let resolver = partition_values
+                    .iter()
+                    .filter(|(k, _v)| self.schema.field(k).is_some())
+                    .map(|(k, v)| {
+                        // The schema we are evaluating only contains the fields of interest,
+                        // not all partition fields. The unwrap is safe because we have already
+                        // checked in the filter above.
+                        let data_type = self.schema.field(k).unwrap().data_type();
+
+                        let DataType::Primitive(primitive_type) = data_type else {
+                            return Err(crate::Error::unsupported(
+                                format!("Partition filtering only supported for primitive types. Found type: {}", data_type)
+                            ));
+                        };
+
+                        let scalar = primitive_type.parse_scalar(v)?;
+                        Ok((ColumnName::new([k]), scalar))
+                    })
+                    .collect::<DeltaResult<HashMap<ColumnName, Scalar>>>()?;
+
+                let filter = DefaultPredicateEvaluator::from(resolver);
+                Ok(filter.eval_expr(&self.predicate, false).unwrap_or(false))
+            });
+
+            let val = match val {
+                Some(Ok(v)) => v,
+                Some(Err(e)) => return Err(e),
+                None => true,
+            };
+
+            self.selection_vector.push(val);
```

Comment:
nit:
```suggestion
            self.selection_vector.push(val.transpose()?.unwrap_or(true));
```
(but most likely all this code will change once we use an input selection vector to skip unwanted rows, which allows us to correctly treat the `add.partitionValues` column as non-nullable)

Reply:
applied